pkg/safe/state.go

```diff
@@ -87,6 +84,86 @@ func StateList[T resource.Resource](ctx context.Context, st state.State, ptr res
 	return NewList[T](got), nil
 }
 
+// WrappedStateEvent holds a state.Event that can be cast to its original Resource type when accessed with Event().
+type WrappedStateEvent[R resource.Resource] struct {
+	event state.Event
+}
+
+func (wse *WrappedStateEvent[T]) Event() (*state.TypedEvent[T], error) {
+	var result state.TypedEvent[T]
+
+	ok := false
+
+	if wse.event.Resource != nil {
+		result.Resource, ok = wse.event.Resource.(T)
+		if !ok {
+			return nil, typeMismatchErr(result.Resource, wse.event.Resource)
+		}
+	}
+
+	if wse.event.Old != nil {
+		result.Old, ok = wse.event.Old.(T)
+		if !ok {
+			return nil, typeMismatchErr(result.Old, wse.event.Old)
+		}
+	}
+
+	return &result, nil
+}
+
+func watch[T resource.Resource](ctx context.Context, eventCh chan<- WrappedStateEvent[T], untypedEventCh <-chan state.Event) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case event := <-untypedEventCh:
+			if resource.IsTombstone(event.Resource) {
```

Comment:
another option is to add getters for `Old()` and `Resource()` to the WrappedStateEvent, and this way we can pass through tombstones, they just won't be convertible. 

this way we don't need `TypedEvent` at all?